src/SAML2/XML/saml/NameIDType.php

```diff
@@ -146,12 +144,33 @@ public function getSPProvidedID(): ?string
      * @param string|null $spProvidedID
      * @return void
      */
-    public function setSPProvidedID(string $spProvidedID = null): void
+    public function setSPProvidedID(?string $spProvidedID): void
     {
         $this->SPProvidedID = $spProvidedID;
     }
 
 
+    /**
+     * Convert XML into an NameIDType
+     *
+     * @param \DOMElement $xml The XML element we should load
+     *
+     * @return \SAML2\XML\saml\NameIDType
+     * @throws \InvalidArgumentException
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        Assert::NotEmpty($xml->textContent);
+
+        $Format = self::getAttribute($xml, 'Format', null);
+        $SPProvidedID = self::getAttribute($xml, 'SPProvidedID', null);
+        $NameQualifier = self::getAttribute($xml, 'NameQualifier', null);
+        $SPNameQualifier = self::getAttribute($xml, 'SPNameQualifier', null);
+
+        return new static($xml->textContent, $Format, $SPProvidedID, $NameQualifier, $SPNameQualifier);
```

Comment:
This is tricky (been there, done that). Any class extending `NameIDType` needs to respect this exact signature of the constructor. We should document this in the phpdoc.

Reply:
I think in this case it is fairly safe, but I don't mind duplicating four lines across Issuer/NameID as well... What would you prefer?

Reply:
I decided to go safe and get rid of the static call at the cost of 4~5 lines code duplication..